applications/imaging_ai_segmentator/monai_totalseg_operator.py

```diff
@@ -0,0 +1,283 @@
+        self._input_dataset_key = "image"
+        self._pred_dataset_key = "pred"
+
+        self.model_path = self._find_model_file_path(model_path)
+        self.output_folder = output_folder
+        self.output_folder.mkdir(parents=True, exist_ok=True)
+        self.app_context = app_context
+        self.input_name_image = "image"
+        self.output_name_seg = "seg_image"
+        self.output_name_saved_images_folder = "saved_images_folder"
+
+        # Call the base class __init__() last.
+        # Also, the base class has an attribute called fragment for storing the fragment object
+        super().__init__(fragment, *args, **kwargs)
+
+    def _find_model_file_path(self, model_path: Path):
+        # We'd just have to find the first file and return it for now,
+        # because it is known that in Holoscan App Package, the path is the root models folder
+        if model_path:
+            if model_path.is_file():
+                return model_path
+            elif model_path.is_dir():
+                for file in model_path.rglob("*"):
+                    if file.is_file():
+                        return file
+
+        raise ValueError(f"Model file not found in the provided path: {model_path}")
+
+    def setup(self, spec: OperatorSpec):
+        spec.input(self.input_name_image)
+        spec.output(self.output_name_seg)
+        spec.output(self.output_name_saved_images_folder).condition(
+            ConditionType.NONE
+        )  # Output not requiring a receiver
+
+    def compute(self, op_input, op_output, context):
+        input_image = op_input.receive(self.input_name_image)
+        if not input_image:
+            raise ValueError("Input image is not found.")
+
+        # This operator gets an in-memory Image object, so a specialized ImageReader is needed.
+        _reader = InMemImageReader(input_image)
+
+        # In this example, the input image, once loaded at the beginning of the pre-transforms, can
+        # be saved on disk, so can the segmentation prediction image at the end of the post-transform.
+        # They are both saved in the same subfolder of the application output folder, with names
+        # distinguished by the postfix. They can also be saved in different subfolder if need be.
+        # These images files can then be packaged for rendering.
+        # In the code below, saving of the image files are disabled to save 10 seconds if nii, and 20 if nii.gz
+        pre_transforms = self.pre_process(_reader, str(self.output_folder))
+        post_transforms = self.post_process(pre_transforms, str(self.output_folder))
+
+        # Load the PyTorch model directly as the model factory limits the support to TorchScript as of now
+
+        _device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        _blocks_down: tuple = (1, 2, 2, 4)
+        _blocks_up: tuple = (1, 1, 1)
+
+        # Create the model and push to device
+        model = monai.networks.nets.segresnet.SegResNet(
```

Comment:
Could we move model setup out of `compute()` to avoid setup on every operator tick? Or is there an underlying assumption that this operator will only run once in this example?

Reply:
Container workload, run once only, at least for now.